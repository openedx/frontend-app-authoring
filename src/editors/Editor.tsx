// Note: there is no Editor.test.tsx. This component only works together with
// <EditorPage> as its parent, so they are tested together in EditorPage.test.tsx
import React from 'react';
import { useDispatch } from 'react-redux';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import messages from './messages';
import * as hooks from './hooks';

import supportedEditors from './supportedEditors';
import type { EditorComponent } from './EditorComponent';
import { useEditorContext } from './EditorContext';

export interface Props extends EditorComponent {
  blockType: string;
  blockId: string | null;
  learningContextId: string | null;
  lmsEndpointUrl: string | null;
  studioEndpointUrl: string | null;
  fullScreen?: boolean;
}

const Editor: React.FC<Props> = ({
  learningContextId,
  blockType,
  blockId,
  lmsEndpointUrl,
  studioEndpointUrl,
  onClose = null,
  returnFunction = null,
}) => {
  const dispatch = useDispatch();
  hooks.initializeApp({
    dispatch,
    data: {
      blockId,
      blockType,
      learningContextId,
      lmsEndpointUrl,
      studioEndpointUrl,
    },
  });
  const { fullScreen } = useEditorContext();

  const EditorComponent = supportedEditors[blockType];
  const innerEditor = (EditorComponent !== undefined)
    ? <EditorComponent {...{ onClose, returnFunction }} />
    : <FormattedMessage {...messages.couldNotFindEditor} />;

  if (fullScreen) {
    return (
      <div
        className="d-flex flex-column"
      >
        <div
          className="pgn__modal-fullscreen h-100"
          role="dialog"
          aria-label={blockType}
        >
          {innerEditor}
        </div>
      </div>
    );
  }
  return innerEditor;
};

export default Editor;
