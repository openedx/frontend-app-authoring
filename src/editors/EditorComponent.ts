/** Shared interface that all Editor components (like ProblemEditor) adhere to */
export interface EditorComponent {
  onClose: (() => void) | null;
  // TODO: get a better type for the 'result' here
  returnFunction?: (() => (result: any) => void) | null;
}
