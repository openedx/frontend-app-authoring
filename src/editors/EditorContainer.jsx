import React from 'react';
import PropTypes from 'prop-types';
import { useParams } from 'react-router';
import { EditorPage } from '@edx/frontend-lib-content-components';

const EditorContainer = ({
  courseId,
}) => {
  const { blockType, blockId } = useParams();
  return (
    <div className="editor-page">
      <EditorPage
        courseId={courseId}
        blockType={blockType}
        blockId={blockId}
        studioEndpointUrl={process.env.STUDIO_BASE_URL}
        lmsEndpointUrl={process.env.LMS_BASE_URL}
      />
    </div>
  );
};
EditorContainer.propTypes = {
  courseId: PropTypes.string.isRequired,
};

export default EditorContainer;
