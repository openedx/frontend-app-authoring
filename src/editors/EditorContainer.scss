.editor-page {
  .pgn__modal-body {
    @include media-breakpoint-down(xs) {
      padding-left: 10px;
      padding-right: 10px;
    }
  }

  .pgn-transition-replace-group .bg-light-300 {
    @include media-breakpoint-down(xs) {
      padding: 5px !important;
    }
  }

  .pgn__form-control-set .pgn__action-row {
    @include media-breakpoint-down(xs) {
      flex-flow: column wrap;
      align-items: flex-start;
      justify-content: flex-start;
      gap: 8px;
    }
  }
}

.pgn__modal.pgn__modal-scroll-fullscreen {
  @include media-breakpoint-down(xs) {
    margin: 0;
  }
}
