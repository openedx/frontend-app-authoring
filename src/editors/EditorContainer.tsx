import React from 'react';
import { useLocation, useParams, useSearchParams } from 'react-router-dom';
import { getConfig } from '@edx/frontend-platform';
import { useIntl } from '@edx/frontend-platform/i18n';
import { Button, Hyperlink } from '@openedx/paragon';
import { Warning as WarningIcon } from '@openedx/paragon/icons';

import EditorPage from './EditorPage';
import AlertMessage from '../generic/alert-message';
import messages from './messages';
import { getLibraryId } from '../generic/key-utils';
import { createCorrectInternalRoute } from '../utils';

interface Props {
  /** Course ID or Library ID */
  learningContextId: string;
  /** Event handler sometimes called when user cancels out of the editor page */
  onClose?: (prevPath?: string) => void;
  /**
   * Event handler called after when user saves their changes using an editor
   * and sometimes called when user cancels the editor, instead of onClose.
   * If changes are saved, newData will be present, and if it was cancellation,
   * newData will be undefined.
   * TODO: clean this up so there are separate onCancel and onSave callbacks,
   * and they are used consistently instead of this mess.
   */
  returnFunction?: (prevPath?: string) => (newData: Record<string, any> | undefined) => void;
}

const EditorContainer: React.FC<Props> = ({
  learningContextId,
  onClose,
  returnFunction,
}) => {
  const intl = useIntl();
  const { blockType, blockId } = useParams();
  const location = useLocation();
  const [searchParams] = useSearchParams();
  const upstreamLibRef = searchParams.get('upstreamLibRef');

  if (blockType === undefined || blockId === undefined) {
    // istanbul ignore next - This shouldn't be possible; it's just here to satisfy the type checker.
    return <div>Error: missing URL parameters</div>;
  }

  const getLibraryBlockUrl = () => {
    if (!upstreamLibRef) {
      return '';
    }
    const libId = getLibraryId(upstreamLibRef);
    return createCorrectInternalRoute(`/library/${libId}/components?usageKey=${upstreamLibRef}`);
  };

  return (
    <div className="editor-page">
      <AlertMessage
        className="m-3"
        show={upstreamLibRef}
        variant="warning"
        icon={WarningIcon}
        title={intl.formatMessage(messages.libraryBlockEditWarningTitle)}
        description={intl.formatMessage(messages.libraryBlockEditWarningDescription)}
        actions={[
          <Button
            destination={getLibraryBlockUrl()}
            target="_blank"
            rel="noopener noreferrer"
            showLaunchIcon
            as={Hyperlink}
          >
            {intl.formatMessage(messages.libraryBlockEditWarningLink)}
          </Button>,
        ]}
      />
      <EditorPage
        courseId={learningContextId}
        blockType={blockType}
        blockId={blockId}
        studioEndpointUrl={getConfig().STUDIO_BASE_URL}
        lmsEndpointUrl={getConfig().LMS_BASE_URL}
        onClose={onClose ? () => onClose(location.state?.from) : null}
        returnFunction={returnFunction ? () => returnFunction(location.state?.from) : null}
      />
    </div>
  );
};

export default EditorContainer;
