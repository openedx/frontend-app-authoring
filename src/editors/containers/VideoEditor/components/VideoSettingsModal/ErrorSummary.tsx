import React from 'react';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import { Alert } from '@openedx/paragon';
import { InfoOutline } from '@openedx/paragon/icons';

import messages from './components/messages';
import { ErrorContext } from '../../hooks';

export const hasNoError = (error) => Object.keys(error[0]).length === 0;

export const showAlert = (errors) => !Object.values(errors).every(hasNoError);

export const ErrorSummary = () => {
  const errors = React.useContext(ErrorContext);
  return (
    <Alert
      icon={InfoOutline}
      show={showAlert(errors)}
      variant="danger"
    >
      <Alert.Heading>
        <FormattedMessage {...messages.validateErrorTitle} />
      </Alert.Heading>
      <p>
        <FormattedMessage {...messages.validateErrorBody} />
      </p>
    </Alert>
  );
};
