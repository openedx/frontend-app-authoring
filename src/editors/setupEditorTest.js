// These additional mocks and setup are required for some tests in src/editors/
// and are imported on an as-needed basis.

// /////////////////////////////////////////////////////////////////////////////
// TODO: tests using this 'setupEditorTest', shallow rendering, and snapshots
// should be replaced with modern tests that use src/testUtils.ts and
// @testing-library/react. See
//   src/editors/containers/ProblemEditor/components/SelectTypeModal/index.test.tsx
// for an example of an editor test in the new format.
// /////////////////////////////////////////////////////////////////////////////

import { formatMessage as mockFormatMessage } from './testUtils';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-canvas-mock';

jest.mock('@edx/frontend-platform/i18n', () => {
  const i18n = jest.requireActual('@edx/frontend-platform/i18n');
  const PropTypes = jest.requireActual('prop-types');
  return {
    ...i18n,
    useIntl: () => ({ formatMessage: mockFormatMessage }),
    intlShape: PropTypes.shape({
      formatMessage: PropTypes.func,
    }),
    defineMessages: m => m,
    getLocale: () => 'getLocale',
    FormattedDate: () => 'FormattedDate',
    FormattedMessage: () => 'FormattedMessage',
    FormattedTime: () => 'FormattedTime',
  };
});

jest.mock('@openedx/paragon', () => jest.requireActual('./testUtils').mockNestedComponents({
  Alert: {
    Heading: 'Alert.Heading',
  },
  AlertModal: 'AlertModal',
  ActionRow: {
    Spacer: 'ActionRow.Spacer',
  },
  Badge: 'Badge',
  Button: 'Button',
  ButtonGroup: 'ButtonGroup',
  Collapsible: {
    Advanced: 'Advanced',
    Body: 'Body',
    Trigger: 'Trigger',
    Visible: 'Visible',
  },
  Card: {
    Header: 'Card.Header',
    Section: 'Card.Section',
    Footer: 'Card.Footer',
    Body: 'Card.Body',
  },
  CheckboxControl: 'CheckboxControl',
  Col: 'Col',
  Container: 'Container',
  Dropdown: {
    Item: 'Dropdown.Item',
    Menu: 'Dropdown.Menu',
    Toggle: 'Dropdown.Toggle',
  },
  ErrorContext: {
    Provider: 'ErrorContext.Provider',
  },
  Hyperlink: 'Hyperlink',
  Icon: 'Icon',
  IconButton: 'IconButton',
  IconButtonWithTooltip: 'IconButtonWithTooltip',
  Image: 'Image',
  MailtoLink: 'MailtoLink',
  ModalDialog: {
    Footer: 'ModalDialog.Footer',
    Header: 'ModalDialog.Header',
    Title: 'ModalDialog.Title',
    Body: 'ModalDialog.Body',
    CloseButton: 'ModalDialog.CloseButton',
  },
  Form: {
    Checkbox: 'Form.Checkbox',
    Control: {
      Feedback: 'Form.Control.Feedback',
    },
    Group: 'Form.Group',
    Label: 'Form.Label',
    Text: 'Form.Text',
    Row: 'Form.Row',
    Radio: 'Radio',
    RadioSet: 'RadioSet',
  },
  OverlayTrigger: 'OverlayTrigger',
  Tooltip: 'Tooltip',
  FullscreenModal: 'FullscreenModal',
  Row: 'Row',
  Scrollable: 'Scrollable',
  SelectableBox: {
    Set: 'SelectableBox.Set',
  },

  Spinner: 'Spinner',
  Stack: 'Stack',
  Toast: 'Toast',
  Truncate: 'Truncate',
  useWindowSize: { height: '500px' },
}));

jest.mock('@openedx/paragon/icons', () => ({
  Close: jest.fn().mockName('icons.Close'),
  Edit: jest.fn().mockName('icons.Edit'),
  Locked: jest.fn().mockName('icons.Locked'),
  Unlocked: jest.fn().mockName('icons.Unlocked'),
}));

// Mock react-redux hooks
// unmock for integration tests
jest.mock('react-redux', () => {
  const dispatch = jest.fn((...args) => ({ dispatch: args })).mockName('react-redux.dispatch');
  return {
    connect: (mapStateToProps, mapDispatchToProps) => (component) => ({
      mapStateToProps,
      mapDispatchToProps,
      component,
    }),
    useDispatch: jest.fn(() => dispatch),
    useSelector: jest.fn((selector) => ({ useSelector: selector })),
  };
});

// Mock the plugins repo so jest will stop complaining about ES6 syntax
jest.mock('frontend-components-tinymce-advanced-plugins', () => ({
  a11ycheckerCss: '',
}));
