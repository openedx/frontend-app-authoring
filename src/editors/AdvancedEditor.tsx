import React, { useEffect } from 'react';
import { getConfig } from '@edx/frontend-platform';
import { useIntl } from '@edx/frontend-platform/i18n';

import { LibraryBlock } from '../library-authoring/LibraryBlock';
import { EditorModalWrapper } from './containers/EditorContainer';
import { ToastContext } from '../generic/toast-context';
import messages from './messages';

interface AdvancedEditorProps {
  usageKey: string,
  onClose: Function | null,
}

const AdvancedEditor = ({ usageKey, onClose }: AdvancedEditorProps) => {
  const intl = useIntl();
  const { showToast } = React.useContext(ToastContext);

  useEffect(() => {
    const handleIframeMessage = (event) => {
      if (event.origin !== getConfig().STUDIO_BASE_URL) {
        return;
      }

      if (event.data.type === 'xblock-event') {
        const { eventName, data } = event.data;

        if (onClose && (eventName === 'cancel'
          || (eventName === 'save' && data.state === 'end'))
        ) {
          onClose();
        } else if (eventName === 'error') {
          showToast(intl.formatMessage(messages.advancedEditorGenericError));
        }
      }
    };

    window.addEventListener('message', handleIframeMessage);

    return () => {
      window.removeEventListener('message', handleIframeMessage);
    };
  }, []);

  return (
    <EditorModalWrapper onClose={onClose as () => void}>
      <LibraryBlock
        usageKey={usageKey}
        view="studio_view"
      />
    </EditorModalWrapper>
  );
};

export default AdvancedEditor;
