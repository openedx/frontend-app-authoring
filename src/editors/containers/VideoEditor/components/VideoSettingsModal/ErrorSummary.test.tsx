import 'CourseAuthoring/editors/setupEditorTest';
import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';

import * as mod from './ErrorSummary';

const { ErrorSummary } = mod;

describe('ErrorSummary', () => {
  const errors = {
    widgetWithError: [{ err1: 'mSg', err2: 'msG2' }, jest.fn()],
    widgetWithNoError: [{}, jest.fn()],
  };
  afterEach(() => {
    jest.restoreAllMocks();
  });
  describe('render', () => {
    beforeEach(() => {
      jest.spyOn(React, 'useContext').mockReturnValueOnce({});
    });
    test('snapshots: renders as expected when there are no errors', () => {
      jest.spyOn(mod, 'showAlert').mockReturnValue(false);
      expect(shallow(<ErrorSummary />).snapshot).toMatchSnapshot();
    });
    test('snapshots: renders as expected when there are errors', () => {
      jest.spyOn(mod, 'showAlert').mockReturnValue(true);
      expect(shallow(<ErrorSummary />).snapshot).toMatchSnapshot();
    });
  });
  describe('hasNoError', () => {
    it('returns true', () => {
      expect(mod.hasNoError(errors.widgetWithError)).toEqual(false);
    });
    it('returns false', () => {
      expect(mod.hasNoError(errors.widgetWithNoError)).toEqual(true);
    });
  });
  describe('showAlert', () => {
    it('returns true', () => {
      jest.spyOn(mod, 'hasNoError').mockReturnValue(false);
      expect(mod.showAlert(errors)).toEqual(true);
    });
    it('returns false', () => {
      jest.spyOn(mod, 'hasNoError').mockReturnValue(true);
      expect(mod.showAlert(errors)).toEqual(false);
    });
  });
});
