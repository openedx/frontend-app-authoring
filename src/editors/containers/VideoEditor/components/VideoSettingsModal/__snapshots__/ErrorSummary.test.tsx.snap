// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorSummary render snapshots: renders as expected when there are errors 1`] = `
<Alert
  show={true}
  variant="danger"
>
  <Alert.Heading>
    <FormattedMessage
      defaultMessage="We couldn't add your video."
      description="Title of validation error."
      id="authoring.videoeditor.validate.error.title"
    />
  </Alert.Heading>
  <p>
    <FormattedMessage
      defaultMessage="Please check your entries and try again."
      description="Body of validation error."
      id="authoring.videoeditor.validate.error.body"
    />
  </p>
</Alert>
`;

exports[`ErrorSummary render snapshots: renders as expected when there are no errors 1`] = `
<Alert
  show={false}
  variant="danger"
>
  <Alert.Heading>
    <FormattedMessage
      defaultMessage="We couldn't add your video."
      description="Title of validation error."
      id="authoring.videoeditor.validate.error.title"
    />
  </Alert.Heading>
  <p>
    <FormattedMessage
      defaultMessage="Please check your entries and try again."
      description="Body of validation error."
      id="authoring.videoeditor.validate.error.body"
    />
  </p>
</Alert>
`;
